src/instruction/instructions/shift_rotate/asrs.ts

```diff
@@ -0,0 +1,126 @@
+import { IMemory } from 'board/memory/interfaces'
+import { Registers } from 'board/registers'
+import { ILabelOffsets } from 'instruction/interfaces'
+import {
+  checkOptionCount,
+  create,
+  createImmediateBits,
+  createLowRegisterBits,
+  getBits,
+  isImmediate,
+  isOptionCountValid,
+  setBits
+} from 'instruction/opcode'
+import { Halfword, Word } from 'types/binary'
+import { BaseInstruction } from '../base'
+
+export class AsrsRegisterInstruction extends BaseInstruction {
+  public name: string = 'ASRS'
+  public pattern: string = '0100000100XXXXXX'
+  private rdnPattern: string = '0100000100000XXX'
+  private rmPattern: string = '0100000100XXX000'
+  private expectedOptionCount: number = 3
+
+  public encodeInstruction(options: string[], labels: ILabelOffsets): Halfword {
+    checkOptionCount(options, this.expectedOptionCount)
+    if (options[0] !== options[1])
+      throw new Error('Parameter 1 and 2 must be identical!')
+    let opcode: Halfword = create(this.pattern)
+    opcode = setBits(opcode, this.rdnPattern, createLowRegisterBits(options[0]))
+    opcode = setBits(opcode, this.rmPattern, createLowRegisterBits(options[2]))
+    return opcode
+  }
+
+  public executeInstruction(
+    opcode: Halfword,
+    registers: Registers,
+    memory: IMemory
+  ): void {
+    let rdnBits = getBits(opcode, this.rdnPattern)
+    let rmBits = getBits(opcode, this.rmPattern)
+    let rdnValue: Word = registers.readRegister(rdnBits.value)
+    let rmValue: Word = registers.readRegister(rmBits.value)
+
+    let result: Word
+    let isCarrySet: boolean = false
+
+    if (rmValue.value < Word.NUMBER_OF_BITS) {
+      result = Word.fromSignedInteger(rdnValue.value >> rmValue.value)
+      isCarrySet =
+        rmValue.value > 0 ? rdnValue.isBitSet(rmValue.value - 1) : false
+    } else {
+      result = rdnValue.hasSign()
+        ? Word.fromUnsignedInteger(Word.MAX_VALUE)
+        : Word.fromUnsignedInteger(Word.MIN_VALUE)
+      isCarrySet = result.hasSign()
+    }
+
+    registers.writeRegister(rdnBits.value, result)
+
+    registers.setFlags({
+      N: result.hasSign(),
+      Z: result.value === 0,
```

Comment:
For N and Z you can use the method `evaluateZeroAndNegativeFlags()` from the ALU component. Same applies for all the other shift instructions.